### PR TITLE
NodeGraph: Use values from fixedX/fixedY column for layout

### LINKED ISSUE #86527: NodeGraph: Allow specifying fixed X and Y coordinates for nodes in DataFrame
**Why is this needed**:

I want to visualize data and statistics for nodes in an existing graph from another application where the user directly places nodes and their connections on a canvas. As such, the nodes I query from the database (postgresql) have X and Y coordinates available, and in order for the user to more easily recognize which nodes from that editor correspond to which nodes in grafana, it would be great to just take the positions from a dataframe. The layout algorithm makes rather a mess of things :)

**What would you like to be added**:

Fields in the DataFrame that would be picked up before the layouting algorithm runs. They could be named x/y, or fixedX/fixedY or staticX/staticY or something similar.

**Who is this feature for?**

Anyone who takes node graph data from a source that has positions already assigned, and anyone who wants to do their own layouting in the query (don't laugh, you can do incredible things in sql ;) )

**Additional notes**

I actually already have a "minimum viable patch" for this, but it would 1) need to be cleaned up and 2) I would like to find out first, if this is even something the grafana team would accept into upstream.

Without coordinates from data:
image
With coordinates from data:
image

You can see that the layout used here is not one that the layouter would ever come up with, what with nodes overlapping and everything (mostly just needs a `y * 2` in the sql query), but the ability to recognize which node is which is very valuable for this use case!

### PR DESCRIPTION
**Pull request by timo**
**What is this feature?**

Allows a DataFrame for the "nodes" part of a NodeGraph panel's data to have a column named fixedX and a column named fixedY that (when present and filled only with non-null/defined/not-NaN values) will be used to place the nodes, instead of running a layout algorithm.

**Why do we need this feature?**

The issue #86527 contains discussion about this. It's for when the nodes come from somewhere that already has a layout. One example is the Thingsboard Rule Chain Editor, where the user places nodes manually.

**Which issue(s) does this PR fix?**:

Fixes #86527

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by timo** [comment]
Todo:

- [ ] add to the tests

**Comment by timo** [comment]
If you would like to test this feature, here's a github supply copypaste thingie that creates a NodeGraph panel that has manually-placed nodes in it, and in the data sources you can use the "hide" toggle to switch between a nodes frame that has fixedX and fixedY and one that doesn't.

| Key | Value |
|--|--|
| Panel | nodeGraph @ Grafana v11.1.0-pre (00db3572dc) |
| Grafana | Grafana v11.1.0-pre (00db3572dc) // Open Source |
<details><summary>Panel debug snapshot dashboard</summary>

```json
{
  "panels": [
    {
      "datasource": {
        "type": "grafana",
        "uid": "grafana"
      },
      "gridPos": {
        "h": 13,
        "w": 15,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "nodes": {},
        "edges": {}
      },
      "targets": [
        {
          "refId": "A",
          "datasource": {
            "type": "grafana",
            "uid": "grafana"
          },
          "queryType": "snapshot",
          "snapshot": [
            {
              "schema": {
                "refId": "nodes",
                "fields": [
                  {
                    "name": "id",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "name",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "title",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "fixedX",
                    "type": "number",
                    "typeInfo": {
                      "frame": "int64",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "fixedY",
                    "type": "number",
                    "typeInfo": {
                      "frame": "int64",
                      "nullable": true
                    },
                    "config": {}
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    "a",
                    "b",
                    "c",
                    "d",
                    "e"
                  ],
                  [
                    "hello",
                    "goodbye",
                    "wow",
                    "beep",
                    "boop"
                  ],
                  [
                    "right",
                    "root",
                    "basement",
                    "beep",
                    "boop"
                  ],
                  [
                    300,
                    0,
                    -50,
                    200,
                    100
                  ],
                  [
                    0,
                    0,
                    150,
                    200,
                    -100
                  ]
                ]
              }
            },
            {
              "schema": {
                "refId": "edges",
                "fields": [
                  {
                    "name": "id",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "source",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "target",
                    "type": "string",
                    "typeInfo": {
                      "frame": "string",
                      "nullable": true
                    },
                    "config": {}
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    "ab",
                    "ac",
                    "de"
                  ],
                  [
                    "a",
                    "a",
                    "d"
                  ],
                  [
                    "b",
                    "c",
                    "e"
                  ]
                ]
              }
            }
          ]
        }
      ],
      "title": "Reproduced with embedded data",
      "type": "nodeGraph"
    },
    {
      "gridPos": {
        "h": 7,
        "w": 9,
        "x": 15,
        "y": 0
      },
      "id": 5,
      "options": {
        "content": "<table width=\"100%\">\n    <tr>\n      <th width=\"2%\">Panel</th>\n      <td >nodeGraph @ Grafana v11.1.0-pre (00db3572dc)</td>\n    </tr>\n    <tr>\n      <th>Queries</th>\n      <td>nodes[grafana-testdata-datasource], edges[grafana-testdata-datasource], A[grafana-testdata-datasource]</td>\n    </tr>\n    \n    <tr><th>Data</th><td> 2 frames, 8 fields, 8 rows </td></tr>\n    \n    <tr>\n      <th>Grafana</th>\n      <td>Grafana v11.1.0-pre (00db3572dc) // Open Source</td>\n    </tr>\n  </table>",
        "mode": "html"
      },
      "title": "Debug info",
      "type": "text"
    },
    {
      "id": 6,
      "title": "Original Panel JSON",
      "type": "text",
      "gridPos": {
        "h": 13,
        "w": 9,
        "x": 15,
        "y": 7
      },
      "options": {
        "content": "{\n  \"datasource\": {\n    \"uid\": \"cdj4bzetkgglcc\",\n    \"type\": \"grafana-testdata-datasource\"\n  },\n  \"gridPos\": {\n    \"h\": 21,\n    \"w\": 20,\n    \"x\": 2,\n    \"y\": 0\n  },\n  \"id\": 1,\n  \"options\": {\n    \"nodes\": {},\n    \"edges\": {}\n  },\n  \"targets\": [\n    {\n      \"scenarioId\": \"csv_content\",\n      \"refId\": \"nodes\",\n      \"datasource\": {\n        \"type\": \"grafana-testdata-datasource\",\n        \"uid\": \"cdj4bzetkgglcc\"\n      },\n      \"csvContent\": \"id,name,title,fixedX,fixedY\\na,\\\"hello\\\",\\\"right\\\",300,0\\nb,\\\"goodbye\\\",\\\"root\\\",0,0\\nc,\\\"wow\\\",\\\"basement\\\",-50,150\\nd,\\\"beep\\\",\\\"beep\\\",200,200\\ne,\\\"boop\\\",\\\"boop\\\",100,-100\",\n      \"hide\": false\n    },\n    {\n      \"datasource\": {\n        \"type\": \"grafana-testdata-datasource\",\n        \"uid\": \"cdj4bzetkgglcc\"\n      },\n      \"scenarioId\": \"csv_content\",\n      \"refId\": \"edges\",\n      \"csvContent\": \"id,source,target\\nab,a,b\\nac,a,c\\nde,d,e\"\n    },\n    {\n      \"scenarioId\": \"csv_content\",\n      \"refId\": \"A\",\n      \"datasource\": {\n        \"type\": \"grafana-testdata-datasource\",\n        \"uid\": \"cdj4bzetkgglcc\"\n      },\n      \"csvContent\": \"id,name,title,random_number_one,random_number_two\\na,\\\"hello\\\",\\\"right\\\",300,0\\nb,\\\"goodbye\\\",\\\"root\\\",0,0\\nc,\\\"wow\\\",\\\"basement\\\",-50,150\\nd,\\\"beep\\\",\\\"beep\\\",200,200\\ne,\\\"boop\\\",\\\"boop\\\",100,-100\",\n      \"hide\": true\n    }\n  ],\n  \"title\": \"Panel Title\",\n  \"type\": \"nodeGraph\"\n}",
        "mode": "code",
        "code": {
          "language": "json",
          "showLineNumbers": true,
          "showMiniMap": true
        }
      }
    },
    {
      "id": 3,
      "title": "Data from panel above",
      "type": "table",
      "datasource": {
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "gridPos": {
        "h": 7,
        "w": 15,
        "x": 0,
        "y": 13
      },
      "options": {
        "showTypeIcons": true
      },
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 2,
          "withTransforms": true,
          "refId": "A"
        }
      ]
    }
  ],
  "schemaVersion": 37,
  "title": "Debug: Panel Title // 2024-04-21 19:57:31",
  "tags": [
    "debug",
    "debug-nodeGraph"
  ],
  "time": {
    "from": "2024-04-21T11:57:31.686Z",
    "to": "2024-04-21T17:57:31.686Z"
  }
}
```
</details>

image

File: public/app/plugins/panel/nodeGraph/utils.ts [hunk 1]
```diff
@@ -138,12 +143,36 @@ export function processNodes(
...
 
+  let hasFixed: boolean | undefined = undefined;
+
   if (nodes) {
     const nodeFields = getNodeFields(nodes);
     if (!nodeFields.id) {
       throw new Error('id field is required for nodes data frame.');
     }
 
+    if (!!nodeFields.fixed_x !== !!nodeFields.fixed_y) {
```
Comment (aocenas) [inline public/app/plugins/panel/nodeGraph/utils.ts:154]: Not 100% sure but wouldn't something like this have the same effect:
```
    const hasFixedPositions =
      nodeFields.fixed_x &&
      nodeFields.fixed_x.values.every((v) => Number.isFinite(v)) &&
      nodeFields.fixed_y &&
      nodeFields.fixed_y.values.every((v) => Number.isFinite(v));
    if (!hasFixedPositions) {
      const somePosFilled =
        (nodeFields.fixed_x && nodeFields.fixed_x.values.some((v) => Number.isFinite(v))) ||
        (nodeFields.fixed_y && nodeFields.fixed_y.values.some((v) => Number.isFinite(v)));
      if (somePosFilled) {
        throw new Error('fixed_x and fixed_y values have to all be filled and valid');
      }
    }
```

It skips specific error for field omission but I think we don't need that distinction. Would also be nice to add a test for this.

File: public/app/plugins/panel/nodeGraph/utils.ts [hunk 2]
```diff
@@ -138,12 +143,36 @@ export function processNodes(
     return { nodes: [], edges: [] };
   }
 
+  let hasFixed: boolean | undefined = undefined;
```
Comment (aocenas) [inline public/app/plugins/panel/nodeGraph/utils.ts:146]: For a bit more clarity can we rename this to `hasFixedPositions` or something like that?

File: packages/grafana-data/src/utils/nodeGraph.ts [hunk 3]
```diff
@@ -38,4 +38,10 @@ export enum NodeGraphDataFrameFieldNames {
 
   // Defines the stroke dash array for the edge [edges]. See SVG strokeDasharray definition for syntax.
   strokeDasharray = 'strokedasharray',
+
+  // Supplies a fixed X position for the node to have in the finished graph.
+  fixedX = 'fixedX',
+
+  // Supplies a fixed Y position for the node to have in the finished graph.
+  fixedY = 'fixedY',
```
Comment (aocenas) [inline packages/grafana-data/src/utils/nodeGraph.ts:46]: I am not exactly sure why, I think we have some problems with the whole case insensitivness checks but all other values are lowercase so lets also do the same here `fixedY = 'fixedy'`

File: public/app/plugins/panel/nodeGraph/utils.ts [hunk 4]
```diff
@@ -45,6 +45,8 @@ export function shortenLine(line: Line, sourceNodeRadius: number, targetNodeRadi
 }
 
 export type NodeFields = {
+  fixed_x?: Field;
+  fixed_y?: Field;
```
Comment (aocenas) [inline public/app/plugins/panel/nodeGraph/utils.ts:49]: Why the underscore naming here? I think we should also just keep the same camelCase naming

**Review by aocenas — COMMENTED** [review]
Hey thanks for the PR, again :), left some generic code comments. I think otherwise this looks good but will have to get back to it on Thursday to test it a bit.

**Comment by aocenas** [comment]
The code works nicely so if the comments are addressed we can merge this soon.

**Comment by aocenas** [comment]
Updated the PR with some changes related to my comments and also some additional type fixes that I noticed when I seen betterer complain about type assertions.

**Review by aocenas — APPROVED** [review]

**Comment by timo** [comment]
TYVM @aocenas I kind of got swamped and couldn't address your comments, so thanks for getting my code into shape and merged!

**Comment by aocenas** [comment]
@timo thank you, you did most of the work anyway :)

**Pull request merged**